### DISCUSSION TITLE
FastAPI using Gunicorn is not using all the workers

### DISCUSSION THREAD

**Discussion by ArijitSinghEDA** [fastapi/fastapi#10707] [Questions] [answered]
### First Check

- [x] I added a very descriptive title here.
- [x] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, BackgroundTasks
from pydantic import BaseModel

app = FastAPI()

class Item(BaseModel):
    params: dict


@app.post("/out_method_1")
def func1(item: Item, background_tasks: BackgroundTasks):
    from folder import pycode_1
    background_tasks.add_task(pycode_1, item.params)
    return 1

@app.post("/out_method_2")
def func2(item: Item, background_tasks: BackgroundTasks):
    from folder import pycode_2
    background_tasks.add_task(pycode_2, item.params)
    return 1
```

### Description

I am running the server using `gunicorn main:app --workers 4 --worker-class uvicorn.workers.UvicornWorker --bind 127.0.0.1:8111`

My `pycode_1` and `pycode_2` functions are like this:
```
def pycode_1(params):
    # Rest of the function
```

The code is an example, but I have multiple post requests in my main file. I am hitting my requests through Postman, and my code is running inside a docker (The above command line is just a placeholder), and I am immediately sending another request the moment I get `1` as response. For `out_method_1`, it ran once and then was stuck on "sending request", but for `out_method_2`, it ran twice before getting stuck on "sending request". Another thing I noticed was it was sending the request to their same workers.

My questions are:
1. Why it did not use the other 2 workers?
2. `gunicorn` documentation mentions that with 4-12 workers it can process hundreds of thousands requests per second. Is there some limitation to its performance with FastAPI due to the fact that `gunicorn` follow the WSGI standard and FastAPI the ASGI standard?
3. Is there any way that I can get the response `1` immediately after hitting my requests through Postman, even though I hit the same request multiple time (even if it executes on the same worker)?

*EDIT*
Changed the file to use `BackgroundTasks` instead of `asyncio.ensure_future()`
Changed the `async def` to `def`

### Operating System

Linux

### FastAPI Version

0.104.1

### Pydantic Version

1.9.2

### Python Version

3.10.11

**Comment by jgould22** [comment, marked as answer]
> Why it did not use the other 2 workers?

The request gets picked up by whatever worker is woke by the linux scheduler when a request comings in on a socket. 

>  I am immediately sending another request the moment I get 1 as response. 

If you mean you are doing this manually then you then yeah you are unlikely to hit another worker 

> gunicorn documentation mentions that with 4-12 workers it can process hundreds of thousands requests per second. 

This can be true but it does depend on what those workers are doing

> Is there any way that I can get the response 1 immediately after hitting my requests through Postman, even though I hit the same request multiple time (even if it executes on the same worker)?

I would start by using something like curl to make my requests because Postman does a lot of magic and people don't always know what it is doing

Some other questions.

Why are you importing in the controller? Move that to the top level so you don't need to worry about what side affects the import machinery might be causing.

**Reply by ArijitSinghEDA** [reply]
> Why are you importing in the controller? Move that to the top level so you don't need to worry about what side affects the import machinery might be causing.

The reason is that I have multiple controllers like this for different functionalities, which have further imports which can be time taking. If I place each import to the top level, it can increase the loading time, and my requirement is to give a response as soon as the application is ready. If there is a more feasible way to handle this I am more than happy to implement it.

**Reply by jgould22** [reply]
If these imports are so heavy weight and nested are importing them in async functions and blocking the event loop?

**Reply by ArijitSinghEDA** [reply]
I am not very familiar with the asynchronous functions and their working. I have changed the functions to synchronous functions too (Mentioned in the *EDIT* section). The reason why I made this change was I read on a discussion in this repo that if functions are changed to `def`, then the `BackgroundTasks` puts the functions in a treadpool automatically.

**Reply by ArijitSinghEDA** [reply]
>I would start by using something like curl to make my requests because Postman does a lot of magic and people don't always know what it is doing

I am making my requests using `curl` and it is sending the requests to different workers. Thank you !
